packages/std/src/ado_base/rates.rs

```diff
@@ -0,0 +1,237 @@
+use crate::{
+    ado_contract::ADOContract,
+    amp::{AndrAddr, Recipient},
+    common::{deduct_funds, Funds},
+    error::ContractError,
+    os::aos_querier::AOSQuerier,
+};
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{ensure, has_coins, Coin, Decimal, Deps, Event, Fraction, SubMsg};
+use cw20::Cw20Coin;
+
+#[cw_serde]
+pub struct RatesResponse {
+    pub msgs: Vec<SubMsg>,
+    pub events: Vec<Event>,
+    pub leftover_funds: Funds,
+}
+
+impl Default for RatesResponse {
+    fn default() -> Self {
+        Self {
+            msgs: Vec::new(),
+            events: Vec::new(),
+            leftover_funds: Funds::Native(Coin::default()),
+        }
+    }
+}
+
+#[cw_serde]
+pub enum RatesMessage {
+    SetRate { action: String, rate: Rate },
+    RemoveRate { action: String },
+}
+
+#[cw_serde]
+pub enum RatesQueryMessage {
+    GetRate { action: String },
+}
+
+/// An attribute struct used for any events that involve a payment
+pub struct PaymentAttribute {
+    /// The amount paid
+    pub amount: Coin,
+    /// The address the payment was made to
+    pub receiver: String,
```

Comment:
Do we want to consider having this be a `Recipient` instead? So a user can attach a message? Unsure if that's a secure change.